src/wp-admin/includes/class-wp-upgrader.php

```diff
@@ -947,6 +1005,144 @@ public static function create_lock( $lock_name, $release_timeout = null ) {
+		}
+
+		/*
+		 * Skip any plugin that has "." as its slug.
+		 * A slug of "." will result in a `$src` value ending in a period.
+		 *
+		 * On Windows, this will cause the 'plugins' folder to be moved,
+		 * and will cause a failure when attempting to call `mkdir()`.
+		 */
+		if ( '.' === $args['slug'] ) {
+			return false;
+		}
+
+		$dest_dir = $wp_filesystem->wp_content_dir() . 'upgrade/temp-backup/';
+		// Create the temp-backup directory if it doesn't exist.
+		if ( (
+				! $wp_filesystem->is_dir( $dest_dir )
+				&& ! $wp_filesystem->mkdir( $dest_dir )
+			) || (
+				! $wp_filesystem->is_dir( $dest_dir . $args['dir'] . '/' )
+				&& ! $wp_filesystem->mkdir( $dest_dir . $args['dir'] . '/' )
+			)
+		) {
+			return new WP_Error( 'fs_temp_backup_mkdir', $this->strings['temp_backup_mkdir_failed'] );
+		}
+
+		$src_dir = $wp_filesystem->find_folder( $args['src'] );
+		$src     = trailingslashit( $src_dir ) . $args['slug'];
+		$dest    = $dest_dir . $args['dir'] . '/' . $args['slug'];
+
+		// Delete the temp-backup directory if it already exists.
+		if ( $wp_filesystem->is_dir( $dest ) ) {
+			$wp_filesystem->delete( $dest, true );
+		}
+
+		// Move to the temp-backup directory.
+		if ( ! move_dir( $src, $dest ) ) {
+			return new WP_Error( 'fs_temp_backup_move', $this->strings['temp_backup_move_failed'] );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Restores the plugin/theme from the temp-backup directory.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
+	 *
+	 * @param array|string $args {
+	 *     Array of data for the temp-backup.
+	 *
+	 *     @type string $slug Plugin slug.
+	 *     @type string $src  File path to directory.
+	 *     @type string $dir  Directory name.
+	 * }
+	 *
+	 * @return bool|WP_Error
+	 */
+	public function restore_temp_backup( $args ) {
+		global $wp_filesystem;
+
+		if ( empty( $args['slug'] ) || empty( $args['src'] ) || empty( $args['dir'] ) ) {
+			return false;
+		}
+
+		$src      = $wp_filesystem->wp_content_dir() . 'upgrade/temp-backup/' . $args['dir'] . '/' . $args['slug'];
+		$dest_dir = $wp_filesystem->find_folder( $args['src'] );
+		$dest     = trailingslashit( $dest_dir ) . $args['slug'];
+
+		if ( $wp_filesystem->is_dir( $src ) ) {
+			// Cleanup.
+			if ( $wp_filesystem->is_dir( $dest ) && ! $wp_filesystem->delete( $dest, true ) ) {
+				return new WP_Error( 'fs_temp_backup_delete', $this->strings['temp_backup_restore_failed'] );
+			}
+
+			// Move it.
+			if ( ! move_dir( $src, $dest ) ) {
+				return new WP_Error( 'fs_temp_backup_delete', $this->strings['temp_backup_restore_failed'] );
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * Deletes a temp-backup.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
+	 *
+	 * @param array|string $args {
+	 *     Array of data for the temp-backup.
+	 *
+	 *     @type string $slug Plugin slug.
+	 *     @type string $src  File path to directory.
```

Comment:
Even though it’s not used in the method it is passed to the method. Shouldn’t we include it?

Reply:
It's possible to call the method without passing `src`, so personally I'd err on the side of removing it so as not to have others think it's necessary. Others may have a different view though.

Reply:
I'm good either way.

Reply:
I think `src` should stay in the hash...because whenever core calls that method what is passed is `$options['hook_extra']['temp_backup']` and `src` is **always** set there.  So, even tho it's not used in that method, it's good to document that it's passed.

Reply:
Happy to go with that, let's keep `src` in.